src/i2c.rs

```diff
@@ -101,6 +101,58 @@
 
 use crate::private;
 
+/// I2C error
+pub trait Error: core::fmt::Debug {
+    /// Convert error to a generic I2C error kind
+    ///
+    /// By using this method, I2C errors freely defined by HAL implementations
+    /// can be converted to a set of generic I2C errors upon which generic
+    /// code can act.
+    fn kind(&self) -> ErrorKind;
+}
+
+/// I2C error kind
+///
+/// This represents a common set of I2C operation errors. HAL implementations are
+/// free to define more specific or additional error types. However, by providing
+/// a mapping to these common I2C errors, generic code can still react to them.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// An unspecific bus error occurred
+    Bus,
```

Comment:
IMO this error is the same as `Other` in practice. Any ErrorKind can be an "i2c bus error", because they're all about i2c. I'd remove it.

Reply:
you can have other errors, as permission errots on linux for example. Now, it's better to pit the minimal amount of error as we can't remove errors, bit we can add errors layer if needed.

Reply:
I don't think distinguishing "unknown error somewhat related to the bus" and "unknown error probably not related to the bus" is valuable. In practice, platform-independent drivers can do nothing about them, they'll just pass them up.

(Also, I'd say platform-specific errors like "permission denied" or "usb-to-i2c converter unplugged" shouldn't be listed in ErrorKind, they should always be `Other`, because again, they're not useful. Drivers will just pass them up)

Reply:
I agree

Reply:
> IMO this error is the same as `Other` in practice. Any ErrorKind can be an "i2c bus error", because they're all about i2c. I'd remove it.

No, this is not the same. This is taken straight from the STM32 implementation and defined in an RM as:
```
A bus error is detected when a START or a STOP condition is detected and is not located after a multiple of 9 SCL clock pulses. A START or a STOP condition is detected when a SDA edge occurs while SCL is high.
The bus error flag is set only if the I2C is involved in the transfer as master or addressed slave (i.e not during the address phase in slave mode).
```

IMNSHO the the ErrorKinds should be a superset of all error native peripherals can distinguish.

Reply:
I agree with @therealprof, these are not the same as Bus errors are a defined type of error for I2C.

Reply:
Ah, I see. "bus error" is a thing with a particular meaning. I'm fine with keeping it then, but the docs for it should be more specific. "Unspecific bus error" sounds like it's a "catch-all" for all i2c errors that are not explicitly categorized...

Reply:
I have updated the documentation of the bus error with the information provided here.